### PR TITLE
[Spark Dataset runner] Fix initialization of metrics accumulator on driver

### PR DESCRIPTION
**Pull request by mosche**
Fix initialization of metrics accumulator on driver (#24809).

Attempting to initialization the metrics accumulator on the executor side during evaluation will fail, of course.  The Spark context is not available, but most of all this needs to be done before starting the evaluation. 

This PR fixes all related APIs to prevent similar bugs in the future.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by mosche** [comment]
R: @aromanenko-dev 
R: @echauchot

**Comment by mosche** [comment]
Kind ping @aromanenko-dev

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/SparkStructuredStreamingRunner.java [hunk 1]
```diff
@@ -222,7 +225,7 @@ private static Future<?> runAsync(Runnable task) {
     return future;
   }
 
-  private static @Nullable Runnable stopSparkSession(SparkSession session, boolean isProvided) {
+  private static @Nullable Runnable sparkStopFun(SparkSession session, boolean isProvided) {
```
Comment (aromanenko-dev) [inline runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/SparkStructuredStreamingRunner.java:228]: nit: what `Fun` means in this context?

**Review by aromanenko-dev — APPROVED** [review]
Thanks. LGTM

File: runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/SparkStructuredStreamingRunner.java (hunk 1 above)
Comment (mosche) [reply runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/SparkStructuredStreamingRunner.java:228]: `Function`, I'll rename to `sparkStopFn` 👍

**Pull request merged**